velox/functions/prestosql/aggregates/SetAggregates.cpp

```diff
@@ -383,6 +402,61 @@ class SetUnionAggregate : public SetBaseAggregate<T> {
   }
 };
 
+// Returns the number of distinct non-null values in a group. This is an
+// internal function only used for testing.
+template <typename T>
+class CountDistinctAggregate : public SetAggAggregate<T, true> {
+ public:
+  explicit CountDistinctAggregate(
+      const TypePtr& resultType,
+      const TypePtr& inputType)
+      : SetAggAggregate<T, true>(resultType, false), inputType_{inputType} {}
+
+  using Base = SetAggAggregate<T, true>;
+
+  bool supportsToIntermediate() const override {
+    return false;
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    exec::Aggregate::setAllNulls(groups, indices);
+    for (auto i : indices) {
+      new (groups[i] + Base::offset_)
+          SetAccumulator<T>(inputType_, Base::allocator_);
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    return Base::extractValues(groups, numGroups, result);
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto flatResult = (*result)->as<FlatVector<int64_t>>();
+    flatResult->resize(numGroups);
+
+    uint64_t* rawNulls = exec::Aggregate::getRawNulls(flatResult);
+    for (auto i = 0; i < numGroups; ++i) {
+      auto* group = groups[i];
+      if (Base::isNull(group)) {
+        Base::clearNull(rawNulls, i);
+        flatResult->set(i, 0);
```

Comment:
Want to make sure about this : If its null , we return 0 (shouldnt it be null ?)

Reply:
Yeah, we return 0 for empty groups by intention.